### PR TITLE
isorms: add checks to code, and document

### PR DESCRIPTION
**Pull request by mtorpey**
This adds documentation for the actual constructors `R(Z)MSIsoByTriple` and also adds checks to them to make them safe (there's also an `NC` version).

### CODE REVIEW

File: gap/attributes/isorms.gi [hunk 1]
```diff
@@ -790,6 +790,133 @@ end);
...
+
+  # Check graph isomorphism
+  if not IsPerm(graph_iso) then
+    ErrorNoReturn("Semigroups: RMSIsoByTriple: usage,\n",
+                  "<triple>[1] should be a permutation,");
+  elif LargestMovedPoint(graph_iso) > nrrows + nrcols then
+    ErrorNoReturn("Semigroups: RMSIsoByTriple: usage,\n",
+                  "<triple>[1] should be a permutation on [1 .. ",
+                  nrrows + nrcols, "],");
+  elif Maximum(ListPerm(graph_iso, nrcols)) > nrcols then
```
Comment (james-d-mitchell) [inline gap/attributes/isorms.gi]: `ForAny([1 .. nrcols], x -> x ^ graph_iso > nrcols)` would be better, I think

File: gap/attributes/isorms.gi [hunk 2]
```diff
@@ -790,6 +790,133 @@ end);
...
+                  "<triple>[2] should be an isomorphism from\n",
+                  "the underlying group of <R1> to that of <R2>,");
+  fi;
+
+  # Check map from rows and cols to H
+  if not ForAll(g2_elms_list, x -> x in G2) then
+    ErrorNoReturn("Semigroups: RMSIsoByTriple: usage,\n",
+                  "<triple>[3] should only contain elements from ",
+                  "the underlying group of <R2>,");
+  elif Length(g2_elms_list) <> nrrows + nrcols then
```
Comment (james-d-mitchell) [inline gap/attributes/isorms.gi]: Maybe swap this with the first check, since it should be cheaper?

File: tst/standard/isorms.tst [hunk 3]
```diff
@@ -566,17 +566,110 @@ true
...
+gap> RMSIsoByTriple(R, R, [(), auto, [(), (), ()]]);
+Error, Semigroups: RMSIsoByTriple: usage,
+<triple>[3] should have length equal to
+the number of rows and columns of <R1>,
+gap> RMSIsoByTriple(R, R, [(), auto, g_elms_list]);
+Error, Semigroups: RMSIsoByTriple: usage,
+<triple>[3] does not define an isomorphism,
+gap> g_elms_list := [(), (), (), (), ()];;
+gap> RMSIsoByTriple(R, R, [(), auto, g_elms_list]);
+((), IdentityMapping( SymmetricGroup( [ 1 .. 4 ] ) ), [ (), (), (), (), () ])
```
Comment (james-d-mitchell) [inline tst/standard/isorms.tst]: Can you add a `BruteForceIsoCheck` for every isomorphism actually created? If they don't take too long that is.

File: tst/standard/isorms.tst [hunk 4]
```diff
@@ -566,17 +566,110 @@ true
...
+<triple>[2] should be an isomorphism from
+the underlying group of <R1> to that of <R2>,
+gap> RZMSIsoByTriple(R, R, [(), auto, [42, 43, 44]]);
+Error, Semigroups: RZMSIsoByTriple: usage,
+<triple>[3] should only contain elements from the underlying group of <R2>,
+gap> RZMSIsoByTriple(R, R, [(), auto, [(), (), ()]]);
+Error, Semigroups: RZMSIsoByTriple: usage,
+<triple>[3] should have length equal to
+the number of rows and columns of <R1>,
+gap> RZMSIsoByTriple(R, R, [(), auto, g_elms_list]);
```
Comment (james-d-mitchell) [inline tst/standard/isorms.tst]: And here can you create the same thing using the NC version of the function, and then check that it is not an isomorphism using the `BruteForceIsoCheck`?

File: doc/isorms.xml [hunk 5]
```diff
@@ -54,6 +54,68 @@
...
+      <A>R2</A> defined by <A>triple</A>, which should be a list consisting of
+      the following:
+
+      <List>
+        <Item>
+          <C><A>triple</A>[1]</C> should be a permutation describing an
+          isomorphism from the graph of <A>R1</A> to the graph of <A>R2</A>,
+          i.e. it should satisfy
+          <C>OnDigraphs(RZMSDigraph(<A>R1</A>), <A>triple</A>[1])
+             = RZMSDigraph(<A>R2</A>)</C>.
```
Comment (james-d-mitchell) [inline doc/isorms.xml]: Make it more explicit what this means for a Rees matrix semigroup without 0. I.e. that the associated graph of is always complete bipartite and so `triple[1]` should stabiliser `[1 .. NrColumns(R)]` setwise, and `[NrColumns(R) + 1 .. NrColumns(R) + NrRows(R)]`.

File: doc/isorms.xml [hunk 6]
```diff
@@ -54,6 +54,68 @@
   </ManSection>
 <#/GAPDoc>
 
+<#GAPDoc Label="RMSIsoByTriple">
+  <ManSection>
+    <Oper Name = "RMSIsoByTriple" Arg = "R1, R2, triple"/>
+    <Oper Name = "RZMSIsoByTriple" Arg = "R1, R2, triple"/>
+    <Returns>An isomorphism.</Returns>
+    <Description>
+      If <A>R1</A> and <A>R2</A> are isomorphic Rees 0-matrix semigroups then
```
Comment (james-d-mitchell) [inline doc/isorms.xml]: Regular and over groups.

File: doc/isorms.xml [hunk 7]
```diff
@@ -54,6 +54,68 @@
...
+               BookName="ref"/>).
+        </Item>
+        <Item>
+          <C><A>triple</A>[3]</C> should be a list of elements from the
+          underlying group of <A>R2</A>.  If the
+          <Ref Attr="Matrix" BookName="ref"/> of <A>R1</A> has <M>m</M> columns
+          and <M>n</M> rows, then the list should have length <M>m + n</M>,
+          where the first <M>m</M> entries should correspond to the columns of
+          <A>R1</A>'s matrix, and the last <M>n</M> entries should correspond to
+          the rows.
```
Comment (james-d-mitchell) [inline doc/isorms.xml]: More specific here, refer to Howie's theorem and the fact that `triple[3]{[1 .. m]}`  should consist of the u_i from Theorem 3.4.1 and `triple[3]{[m + 1 ... n]}` consist of the `v_lambda ^ -1`, subject to checking that I haven't got `u` and `v` muddled up.

**Comment by mtorpey** [comment]
@james-d-mitchell: changed as you suggested.

**Pull request merged**